\documentclass[woside,a4paper,12pt]{article}
\usepackage[sc]{mathpazo}
\usepackage[german]{babel}
\usepackage[utf8]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{float}
\usepackage{graphicx}
\usepackage{subcaption}
\usepackage{geometry}
\geometry{verbose,tmargin=2.5cm,bmargin=2.5cm,lmargin=2.5cm,rmargin=2.5cm}
\setcounter{secnumdepth}{2}
\setcounter{tocdepth}{2}
\usepackage{url}
\usepackage[unicode=true,pdfusetitle,
 bookmarks=true,bookmarksnumbered=true,bookmarksopen=true,bookmarksopenlevel=2,
 breaklinks=false,pdfborder={0 0 1},backref=false,colorlinks=false]
 {hyperref}
\hypersetup{pdfstartview={XYZ null null 1}}
\usepackage{breakurl}
\usepackage{booktabs}
\usepackage{longtable}
\usepackage{array}
\usepackage{multirow}
\usepackage[table]{xcolor}
\usepackage{wrapfig}
\usepackage{float}
\usepackage{colortbl}
\usepackage{pdflscape}
\usepackage{tabu}
\usepackage{threeparttable}
\usepackage{fullpage}
\usepackage{pdflscape}

\begin{document}

<<setup, include=FALSE, cache=FALSE, echo=FALSE, eval=TRUE>>=
library(knitr)
library(kableExtra)
options(knitr.table.format = "latex")
# set global chunk options
opts_chunk$set(fig.path = "figure/", fig.align = "center", fig.show = "hold")
options(formatR.arrow = TRUE, width = 90)
load("WES.RData")
sample <- paste(unlist(strsplit(sample, split = "_")), collapse = " ")
@

\title{\Sexpr{sample} - Report}

\author{Patrick Metzger}

\maketitle
\tableofcontents
\clearpage

\section{Qualität der Genomsequenzierung}
\subsection{Raw Quality}

\begin{figure}[H]
  \centering
    \begin{subfigure}[b]{0.45\textwidth}
      \includegraphics[width=\textwidth]{\Sexpr{tumor}}
      \caption{Tumor Qualität}
      \label{fig:1}
    \end{subfigure}
    \begin{subfigure}[b]{0.45\textwidth}
      \includegraphics[width=\textwidth]{\Sexpr{germline}}
      \caption{Keimbahn Qualität}
      \label{fig:2}
    \end{subfigure}
\end{figure}

\subsection{Base Quality Score Recalibration (BQSR)}

\begin{figure}[H]
  \centering
    \begin{subfigure}[b]{0.45\textwidth}
      \includegraphics[width=\textwidth]{\Sexpr{tumor_bsqr}}
      \caption{Tumor Qualität nach BQSR}
      \label{fig:3}
    \end{subfigure}
    \begin{subfigure}[b]{0.45\textwidth}
      \includegraphics[width=\textwidth]{\Sexpr{germline_bsqr}}
      \caption{Keimbahn Qualität nach BQSR}
      \label{fig:4}
    \end{subfigure}
\end{figure}

\subsection{Zusammenfassung}

\begin{itemize}
\item VCRome
\item Paired end 100bp
\item TD: \Sexpr{stats$avreads$nRT} Mio. Reads
\item GD: \Sexpr{stats$avreads$nRG} Mio. Reads
\item Gute Qualität der Reads
\end{itemize}

\section{Coverage}

\begin{figure}[H]
\centering
\includegraphics[width=\textwidth]{\Sexpr{coverage_out}}
\caption{Coverage}
\label{fig:5}
\end{figure}

\subsection{Mean Coverage}
<<echo=FALSE, eval=TRUE>>=
# Mean Coverage
  for (i in 1:length(stats$cover$files)) {
    print(paste("Mean Coverage", stats$cover$labs[i], ":",
                sum(stats$cover$cov[[i]][, 2] * stats$cover$cov[[i]][, 5]),
                sep = " "))
  }
@

\clearpage
\section{Mutationsanalyse}
\subsection{Informationen zur Analyse}

\begin{itemize}
\item Aligned zum Referenzgenom UCSC hg19
\item Einschlusskriterien der Mutation
  \begin{itemize}
    \item Mindestens 8 Reads pro Base
    \item Seltene Mutationen (Minor-Allele Frequency (MAF) $< 0.001$, basierend auf gnomAD exome, ExAC, ESP6500 und 1000g)
    \item Keine \grqq Black-listed\grqq{} Gene/Sequenzen
    \item Variant Allele Frequency (VAF) $> 10\%$
  \end{itemize}
  \item Analyse der Mutationen
  \begin{itemize}
  \item Annotation bekannter Mutationen (Cosmic, Clinvar, dbSNP)
  \item Ranking der Wichtigkeit (RVIS Score)
  \item Strukturanalyse der mutierten Proteine (Condel, CADD)
  \end{itemize}
\end{itemize}

\subsection{Somatische Mutationen und Loss of Heterozygosity (LoH)}

\begin{figure}[H]
\includegraphics[width=\textwidth]{\Sexpr{outfile_circos}}
\caption{Circos Plot}
\label{fig:6}
\end{figure}

\clearpage

<<eval=TRUE, results='tex', echo=FALSE>>=
tmp <- read.delim('MutationTable.txt')
colnames(tmp) <- c("Mutationstype", "Number of exonic", "Zygosity", "TS", "OG", "HS")
kable(tmp, booktabs = T, row.names = F, caption = "Zusammenfassung der identifizierten Mutationen") %>% kable_styling(latex_options = c("striped", "hold_position"), font_size = 12)
n_s <- sum(tmp[c(1:2, 4:5), 2])
n_l <- sum(tmp[c(3,6), 2])
n_a <- n_s + n_l
@

\begin{itemize}
\item \Sexpr{n_s} somatische Mutationen (exonisch)
\item \Sexpr{n_l} Loss of Heterozygosity (LoH)
\item Insgesamt \Sexpr{n_a} Mutationen
\item Mutationslast \Sexpr{round(filt_result_td$tmb, digits = 2)}/Mb
\end{itemize}
<<echo=FALSE, results='tex', eval=TRUE>>=
tmp <- mutation_analysis_result$ts_og
if(dim(tmp)[1]!=0){
tmp <- tmp[,c("Gene.refGene","GeneName","is_tumorsuppressor", "is_oncogene", "is_hotspot")]
colnames(tmp) <- c('Symbol', 'Gene Name','TSG', 'OG', "HS")
kable(tmp, format = "latex", caption = "Tumorsuppressoren und Onkogene - Überblick", booktabs = T, longtable = T,
      digits = 3, row.names = F) %>% kable_styling(latex_options = c("striped", "HOLD_position"), font_size = 8) %>% column_spec(c(1, 3:5), width = "5em") %>% column_spec(2, width = "20em")
  }else{print("No Tumorsuppressors or Oncogenes!")}
@
\clearpage

\begin{landscape}
\subsection{Tumorsuppressoren und Onkogene}
\thispagestyle{empty}
<<echo=FALSE, results='tex', eval=TRUE>>=
tmp <- mutation_analysis_result$ts_og
if(dim(tmp)[1]!=0){
  colnames(tmp) <- c('Symbol', 'Gene Name', 'Exonic Function', 'Aminoacid Change', 'VAF', 'Zygosity', 'Reads', 'TSG', 'OG', 'HS', 'TARGET', 'MAF', 'CADD', 'Condel', 'CLINSIG', 'COSMIC')
  tmp$MAF <- format(tmp$MAF,digits=3, scientific = TRUE)
  tmp$COSMIC <- sub(";.*", "", tmp$COSMIC)
  tmp$COSMIC <- gsub(",", ", ", tmp$COSMIC)
  kable(tmp[,c(1:16)], format = "latex", caption = "Tumorsuppressoren und Onkogene", booktabs = T, digits = 3, row.names = F) %>% kable_styling(latex_options = c("striped", "HOLD_position", "repeat_header"),  font_size = 8) %>% column_spec(c(2:4,10:11), width = "6em") %>% column_spec(c(1,5:9, 12:15), width = "2em") %>% column_spec(16, width = "8em") %>% row_spec(0, angle = 45)
}else {print("No Tumorsuppressors or Oncogenes!")}
@
\clearpage
\subsection{Somatische Mutationen (top20 nach VAF)}
\thispagestyle{empty}
<<echo=FALSE, results='tex', eval=TRUE>>=
tmp <- mutation_analysis_result$som_mut_tab
colnames(tmp) <- c('Symbol', 'Gene Name', 'Exonic Function', 'Aminoacid Change', 'VAF', 'Zygosity', 'Reads',
                     'TSG', 'OG', 'HS', 'TARGET', 'MAF', 'CADD', 'Condel', 'CLINSIG', 'COSMIC')
tmp <- tmp[order(tmp$'VAF', decreasing =  TRUE), , drop = FALSE]
if(dim(tmp)[1]!=0){
  tmp$MAF <- format(tmp$MAF,digits=3, scientific = TRUE)
  tmp$COSMIC <- sub(";.*", "", tmp$COSMIC)
  tmp$COSMIC <- gsub("," , ", ", tmp$COSMIC)
  if(dim(tmp)[1]>20){
    kable(tmp[c(1:20),c(1:16)], format = "latex", caption = "somatische Mutationen", booktabs = T, longtable = T, digits = 3, row.names = F) %>% kable_styling(latex_options = c("striped", "HOLD_position", "repeat_header"), font_size = 8) %>% column_spec(c(1:4,10:11), width = "6em") %>% column_spec(c(5:9, 12:15), width = "2em") %>% column_spec(16, width = "8em") %>% row_spec(0, angle = 45)} else{
     kable(tmp[ ,c(1:16)], format = "latex", caption = "somatische Mutationen", booktabs = T, longtable = T, digits = 3, row.names = F) %>% kable_styling(latex_options = c("striped", "HOLD_position", "repeat_header"), font_size = 8) %>% column_spec(c(1:4,10:11), width = "6em") %>% column_spec(c(5:9, 12:15), width = "2em") %>% column_spec(16, width = "8em") %>% row_spec(0, angle = 45)
    }
}else {print("No somatic mutations!")}
@
\clearpage
\subsection{LoH}
\thispagestyle{empty}
<<echo=FALSE, results='tex', eval=TRUE>>=
if(dim(mutation_analysis_result$table_loh_mutations)[1]>0){
  tmp <- mutation_analysis_result$table_loh_mutations
  colnames(tmp) <- c('Symbol', 'Gene Name', 'Exonic Function', 'Aminoacid Change', 'VAF_Normal','VAF_Tumor','Reads_Normal','Reads_Tumor', 'TSG', 'OG', 'HS', 'TARGET', 'MAF', 'CADD', 'Condel','CLINSIG','COSMIC')
  tmp$MAF <- format(tmp$MAF,digits=3, scientific = TRUE)
  tmp$COSMIC <- sub(";.*", "", tmp$COSMIC)
  tmp$COSMIC <- gsub("," , ", ", tmp$COSMIC)
  kable(tmp[,c(1:17)], format = "latex", caption = "LoH", booktabs = T, longtable = T, digits = 3, row.names = F) %>% kable_styling(latex_options = c("striped", "HOLD_position"), font_size = 8) %>% column_spec(c(1, 3:4, 12), width = "6em") %>% column_spec(c(5:11, 13:16), width = "2em") %>% column_spec(c(2, 17), width = "8em") %>% row_spec(0, angle = 45)
} else {print("No LoH!")}
@
\end{landscape}

\begin{landscape}
\subsection{Funktionelle Analysen}
\subsubsection{GO}
\thispagestyle{empty}
<<echo=FALSE, results='tex', eval=TRUE>>=
if(dim(mutation_analysis_result$go)[1] >= 20){
  tmp <- mutation_analysis_result$go[1:20,c("Term","Count", "Size", "p-value", "adj.P.Val")]
  tmp$`p-value` <- format(tmp$`p-value`,digits=3, scientific = TRUE)
  tmp$adj.P.Val <- format(tmp$adj.P.Val,digits=3, scientific = TRUE)
  kable(tmp, caption = 'Ergebnisse GO Analyse, top 20', booktabs = T, row.names = F) %>% kable_styling(latex_options = c("striped", "HOLD_position"))
} else {
  tmp <- mutation_analysis_result$go[1:dim(mutation_analysis_result$go)[1],c("Term","Count", "Size", "p-value", "adj.P.Val")]
  tmp$`p-value` <- format(tmp$`p-value`,digits=3, scientific = TRUE)
  tmp$adj.P.Val <- format(tmp$adj.P.Val,digits=3, scientific = TRUE)
  kable(tmp, caption = 'Ergebnisse GO Analyse, top 20', booktabs = T, row.names = F) %>% kable_styling(latex_options = c("striped", "HOLD_position"))
}
@
\clearpage
\subsubsection{Consensus}
\thispagestyle{empty}
<<echo=FALSE, results='tex', eval=TRUE>>=
if(dim(mutation_analysis_result$consensus)[1] >= 20){
  tmp <- mutation_analysis_result$consensus[1:20,c("Term","Count", "Size", "p-value", "adj.P.Val","Symbol")]
  tmp$`p-value` <- format(tmp$`p-value`,digits=3, scientific = TRUE)
  tmp$adj.P.Val <- format(tmp$adj.P.Val,digits=3, scientific = TRUE)
  tmp$'Symbol' <- gsub("|", "| ", tmp$'Symbol', fixed=TRUE)
  kable(tmp, format = "latex", caption = 'Ergebnisse Consensus Analyse, top 20', booktabs = T, row.names = F) %>% kable_styling(latex_options = c("striped", "HOLD_position", "repeat_header"),  font_size = 10) %>% column_spec(6, width = "10em")
} else {
  tmp <- mutation_analysis_result$consensus[1:dim(mutation_analysis_result$consensus)[1],c("Term","Count", "Size", "p-value", "adj.P.Val","Symbol")]
  tmp$`p-value` <- format(tmp$`p-value`,digits=3, scientific = TRUE)
  tmp$adj.P.Val <- format(tmp$adj.P.Val,digits=3, scientific = TRUE)
  tmp$'Symbol' <- gsub("|", "| ", tmp$'Symbol', fixed=TRUE)
  kable(tmp, format = "latex", caption = 'Ergebnisse Consensus Analyse, top 20', booktabs = T, row.names = F) %>% kable_styling(latex_options = c("striped", "HOLD_position", "repeat_header"),  font_size = 10) %>% column_spec(6, width = "10em")
}
@

\clearpage
\subsubsection{Reactome}
\thispagestyle{empty}
<<echo=FALSE, results='tex', eval=TRUE>>=
if(dim(mutation_analysis_result$reactome)[1] >= 20){
  tmp <- mutation_analysis_result$reactome[1:20,c("Term","Count", "Size", "p-value", "adj.P.Val")]
  tmp$`p-value` <- format(tmp$`p-value`,digits=3, scientific = TRUE)
  tmp$adj.P.Val <- format(tmp$adj.P.Val,digits=3, scientific = TRUE)
  kable(tmp, caption = 'Ergebnisse Reactome Analyse, top 20', booktabs = T, row.names = F) %>% kable_styling(latex_options = c("striped", "HOLD_position", "scale_down"))
} else {
  tmp <- mutation_analysis_result$reactome[1:dim(mutation_analysis_result$reactome)[1],c("Term","Count", "Size", "p-value", "adj.P.Val")]
  tmp$`p-value` <- format(tmp$`p-value`,digits=3, scientific = TRUE)
  tmp$adj.P.Val <- format(tmp$adj.P.Val,digits=3, scientific = TRUE)
  kable(tmp, caption = 'Ergebnisse Reactome Analyse, top 20', booktabs = T, row.names = F) %>% kable_styling(latex_options = c("striped", "HOLD_position"))
}

@
\clearpage
\subsubsection{Hallmarks of Cancer}
\thispagestyle{empty}
<<echo=FALSE, results='tex', eval=TRUE>>=
if(dim(mutation_analysis_result$hallmarks)[1] >= 20){
  tmp <- mutation_analysis_result$hallmarks[1:20,c("Term","Count", "Size", "p-value", "adj.P.Val")]
  tmp$`p-value` <- format(tmp$`p-value`,digits=3, scientific = TRUE)
  tmp$adj.P.Val <- format(tmp$adj.P.Val,digits=3, scientific = TRUE)
  kable(tmp, caption = 'Ergebnisse Hallmarks of Cancer Analyse, top 20', booktabs = T, row.names = F) %>% kable_styling(latex_options = c("striped", "HOLD_position", "scale_down"))
} else {
  tmp <- mutation_analysis_result$hallmarks[1:dim(mutation_analysis_result$hallmarks)[1],c("Term","Count", "Size", "p-value", "adj.P.Val")]
  tmp$`p-value` <- format(tmp$`p-value`,digits=3, scientific = TRUE)
  tmp$adj.P.Val <- format(tmp$adj.P.Val,digits=3, scientific = TRUE)
  kable(tmp, caption = 'Ergebnisse Hallmarks Analyse, top 20', booktabs = T, row.names = F) %>% kable_styling(latex_options = c("striped", "HOLD_position"))
}

@
\clearpage
\subsection{Wichtige Signalwege}
\thispagestyle{empty}
<<echo=FALSE, eval=TRUE, results='tex'>>=
options(knitr.kable.NA = 'None')
importantPathways <- mutation_analysis_result$important_pathways
if (length(importantPathways) == 0){
  print("Keine Mutationen in wichtigen Signalwegen!")
} else{
  tbl <- importantPathways[ ,c('Pathway', 'Symbol', 'GeneName', 'ExonicFunc', 'VAF', 'Reads', 'AAChange', 'target', 'MAF', 'CADD13', 'Condel', 'CLINSIG', 'COSMIC86')]
  colnames(tbl) <- c('Pathway', 'Symbol', 'Gene Name', 'Exonic Function', 'VAF', 'Reads', 'Aminoacid Change', 'TARGET', 'MAF', 'CADD', 'Condel', 'CLINSIG', 'COSMIC')
  tbl$COSMIC <- sub(";.*", "", tbl$COSMIC)
  tbl$COSMIC <- gsub(",", ", ", tbl$COSMIC)

  kable(tbl, caption = "Important Pathways", booktabs = F, align = "c", longtable = TRUE, row.names = F) %>% kable_styling(latex_options = c("striped", "HOLD_position", "repeat_header"), full_width = FALSE, font_size = 6) %>% column_spec(c(2,4:6, 9:12), width = "3em") %>% column_spec(c(1,3,7), width = "10em") %>% column_spec(c(8,13), width = "15em") %>% column_spec(1, bold = T) %>% row_spec(0, angle = 45)
}
@
\clearpage
\end{landscape}

\section{Copy Number Variation}
\subsection{Anzahl CNVs}
<<echo=FALSE, eval=TRUE>>=
print(paste("#CNVs: ", dim(cnv_analysis_results$cnvs_annotated$CNVsAnnotated)[1], sep = ''))
@
\subsection{CNV Plots}
\begin{figure}[H]
\centering
\includegraphics[width=\textwidth]{\Sexpr{cnv_plot}}
\caption{Copy Number Variation}
\label{fig:11}
\end{figure}

\begin{figure}[H]
\centering
\includegraphics[width=\textwidth]{\Sexpr{cnv_ideogram_plot}}
\caption{Copy Number Variation - Ideogram}
\label{fig:12}
\end{figure}

\subsection{Tumorsuppressoren}
<<echo=FALSE, results='tex', eval=TRUE>>=
kable(cnv_analysis_results$cnvs_annotated$CNVTumorSuppressors, format = "latex", caption = "Tumorsuppressoren", booktabs = T, digits = 3, row.names = F) %>% kable_styling(latex_options = c("striped", "HOLD_position"))
@

\subsection{Onkogene}
<<echo=FALSE, results='tex', eval=TRUE>>=
kable(cnv_analysis_results$cnvs_annotated$CNVOncogenes, format = "latex", caption = "Onkogene", booktabs = T, digits = 3, row.names = F) %>% kable_styling(latex_options = c("striped", "HOLD_position"))
@
\clearpage
\begin{landscape}
\thispagestyle{empty}
\subsection{Funktionelle Analyse der CNVs}
\subsubsection{GAIN}
<<echo=FALSE, results='tex',eval=TRUE>>=
if(dim(cnv_analysis_results$cnv_analysis_results$gain_go)[1] >= 20){
  tmp <- cnv_analysis_results$cnv_analysis_results$gain_go[1:20,c("Term","Count", "Size", "p-value", "adj.P.Val")]
  tmp$`p-value` <- format(tmp$`p-value`,digits=3, scientific = TRUE)
  tmp$adj.P.Val <- format(tmp$adj.P.Val,digits=3, scientific = TRUE)
  kable(tmp, caption = "Ergebnisse GO Analsye - GAIN, top 20", booktabs = T, longtable = T, row.names = F) %>% kable_styling(latex_options = c("striped", "HOLD_position")) %>% column_spec(1, width = "35em") %>% column_spec(c(2:3,5), width = "3em") %>% column_spec(4, width = "5em")
  } else {
  tmp <- cnv_analysis_results$cnv_analysis_results$gain_go[1:dim(cnv_analysis_results$cnv_analysis_results$gain_go)[1],c("Term","Count", "Size", "p-value", "adj.P.Val")]
  tmp$`p-value` <- format(tmp$`p-value`,digits=3, scientific = TRUE)
  tmp$adj.P.Val <- format(tmp$adj.P.Val,digits=3, scientific = TRUE)
  kable(tmp, caption = "Ergebnisse GO Analsye - GAIN, top 20", booktabs = T, longtable = T, row.names = F) %>% kable_styling(latex_options = c("striped", "HOLD_position")) %>% column_spec(1, width = "35em") %>% column_spec(c(2:3,5), width = "3em") %>% column_spec(4, width = "5em")
}
@
\clearpage
\subsubsection{LOSS}
\thispagestyle{empty}
<<echo=FALSE, results='tex',eval=TRUE>>=
if(dim(cnv_analysis_results$cnv_analysis_results$loss_go)[1] >= 20){
  tmp <- cnv_analysis_results$cnv_analysis_results$loss_go[1:20,c("Term","Count", "Size", "p-value", "adj.P.Val")]
  tmp$`p-value` <- format(tmp$`p-value`,digits=3, scientific = TRUE)
  tmp$adj.P.Val <- format(tmp$adj.P.Val,digits=3, scientific = TRUE)
  kable(tmp, caption = "Ergebnisse GO Analyse - LOSS, top 20", booktabs = T, longtable = T, row.names = F) %>% kable_styling(latex_options = c("striped", "HOLD_position")) %>% column_spec(1, width = "35em") %>% column_spec(c(2:3,5), width = "3em") %>% column_spec(4, width = "5em")
} else {
  tmp <- cnv_analysis_results$cnv_analysis_results$loss_go[1:dim(cnv_analysis_results$cnv_analysis_results$loss_go)[1],c("Term","Count", "Size", "p-value", "adj.P.Val")]
  tmp$`p-value` <- format(tmp$`p-value`,digits=3, scientific = TRUE)
  tmp$adj.P.Val <- format(tmp$adj.P.Val,digits=3, scientific = TRUE)
  kable(tmp, caption = "Ergebnisse GO Analyse - LOSS, top 20", booktabs = T, longtable = T, row.names = F) %>% kable_styling(latex_options = c("striped", "HOLD_position")) %>% column_spec(1, width = "35em") %>% column_spec(c(2:3,5), width = "3em") %>% column_spec(4, width = "5em")
}
@
\clearpage
\end{landscape}

\section{Analyse der Mutationssignaturen}

\begin{itemize}
\item Nur somatische Mutationen werden berücksichtigt
\item Nur Signaturen, die mehr als $1\%$ aller SNVs beinhalten, werden verwendet
\item Die Signautren basieren auf den aktuellen \textit{COSMIC Mutation Signatures} \url{http://cancer.sanger.ac.uk/cosmic/signatures}
\item \textit{AC3} wird als \textit{BRCAness} bezeichnet
\end{itemize}

<<echo=FALSE, eval=TRUE, results='tex'>>=
tmp <- mut_sig_analysis$CosmicValid_cutoffGen_LCDlist$norm_exposures
tmp[,1] <- tmp[,1] * 100
tmp <- cbind(mut_sig_analysis$CosmicValid_cutoffGen_LCDlist$out_sig_ind_df,tmp)
tmp <- tmp[,c(1,4,5)]
colnames(tmp) <- c("Signature", "Process", "Percentage")
kable(tmp, caption = "Ergebnisse Mutationssignatur Analyse", booktabs = T, row.names = F, digits=2) %>% kable_styling(latex_options = c("striped", "HOLD_position"))
@

\clearpage

\section{Versionsinfo}

\subsection{Genome}
\begin{itemize}
\item UCSC hg19
\end{itemize}

\subsection{Programmversionen}
\begin{itemize}
\item FastQC: 0.11.5
\item Trimmomatic: 0.38
\item BWA: 0.7.17
\item bam-readcount: 0.8.0
\item samtools 1.9
\item GATK: 3.8.1.0
\item picard-tools: 2.18.15
\item VarScan: 2.4.3
\item annovar 2018-04-16
\item bedtools: 2.27.1
\item Control-FREEC: 11.0
\item Java: 1.8.0\_152
\item R: 3.4.1
\end{itemize}

\subsection{Annovar Datenbanken}
\begin{itemize}
\item refGene GRCh37
\item esp6500siv2\_ea (ESP3500, European)
\item avsnp150 (dbSNP)
\item EU.sites.2015\_08 (1000Genome, European)
\item cosmic86
\item exac03 (ExAC)
\item gnomad\_exome
\item clinvar\_20180603
\item intervar\_20180118
\item dbnsfp35a
\item cadd13
\end{itemize}

\end{document}
